File: Artem.GoogleMap/UI/GooglePolygon.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Runtime.Serialization;
using System.Security.Permissions;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace Artem.Google.UI {

    /// <summary>
    /// Polygon (like a polyline) defines a series of connected coordinates in an ordered sequence; 
    /// additionally, polygons form a closed loop and define a filled region.
    /// </summary>
    [ParseChildren(true, "Points")]
    public class GooglePolygon {

        #region Fields  ///////////////////////////////////////////////////////////////////////////

        List<GoogleLocation> _points;

        #endregion

        #region Properties  ///////////////////////////////////////////////////////////////////////

        #region Appearance

        /// <summary>
        /// The fill color in HTML hex style, ie. "#00AAFF"
        /// </summary>
        /// <value>The color of the fill.</value>
        [Category("Appearance")]
        public Color FillColor { get; set; }

        /// <summary>
        /// The fill opacity between 0.0 and 1.0
        /// </summary>
        /// <value>The fill opacity.</value>
        [Category("Appearance")]
        public float FillOpacity { get; set; }

        /// <summary>
        ///The stroke color in HTML hex style, ie. "#FFAA00"
        /// </summary>
        /// <value>The color of the stroke.</value>
        [Category("Appearance")]
        public Color StrokeColor { get; set; }

        /// <summary>
        /// The stroke opacity between 0.0 and 1.0
        /// </summary>
        /// <value>The stroke opacity.</value>
        [Category("Appearance")]
        public float StrokeOpacity { get; set; }

        /// <summary>
        /// The stroke width in pixels.
        /// </summary>
        /// <value>The stroke weight.</value>
        [Category("Appearance")]
        public int StrokeWeight { get; set; }

        /// <summary>
        /// The zIndex compared to other polys.
        /// </summary>
        /// <value>The index of the Z.</value>
        [Category("Appearance")]
        public int ZIndex { get; set; }

        #endregion

        #region Behavior

        /// <summary>
        /// Indicates whether this Polygon handles click events. Defaults to true.
        /// </summary>
        /// <value>
        /// 	<c>true</c> if this instance is clickable; otherwise, <c>false</c>.
        /// </value>
        [Category("Behavior")]
        public bool IsClickable { get; set; }

        /// <summary>
        /// Render each edge as a geodesic (a segment of a "great circle"). 
        /// A geodesic is the shortest path between two points along the surface of the Earth.
        /// </summary>
        /// <value>
        /// 	<c>true</c> if this instance is geodesic; otherwise, <c>false</c>.
        /// </value>
        [Category("Behavior")]
        public bool IsGeodesic { get; set; }

        #endregion

        #region Data

        /// <summary>
        /// Gets or sets the points.
        /// </summary>
        /// <value>The points.</value>
        [Browsable(true)]
        [Category("Data")]
        [Editor(typeof(CollectionEditor), typeof(UITypeEditor))]
        [PersistenceMode(PersistenceMode.InnerProperty)]
        public List<GoogleLocation> Points {
            get {
                if (_points == null) {
                    _points = new List<GoogleLocation>();
                }
                return _points;
            }
            set {
                _points = value;
            }
        }
        #endregion

        #region Not Yet Supported in GoogleMaps API v3

        /// <summary>
        /// Gets or sets the bounds of the polygon. 
        /// Have in mind it requires an additional post back after polygon is loaded.
        /// </summary>
        /// <value>The bounds.</value>
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        [Obsolete("Not yet available in GoogleMaps API v3")]
        public GoogleBounds Bounds { get; set; }

        /// <summary>
        /// Maps the <c>GPolyEditingOptions</c> <c>fromStart</c> property.
        /// This property specifies whether enableDrawing should add points from the start 
        /// rather than from the end, which is the default.
        /// </summary>
        /// <value>From start.</value>
        [Category("Behavior")]
        [Obsolete("Not yet available in GoogleMaps API v3")]
        public bool? EditingFromStart { get; set; }

        /// <summary>
        /// Maps the <c>GPolyEditingOptions</c> <c>maxVertices</c> property.
        /// This property specifies the maximum number of vertices permitted for this polyline. 
        /// Once this number is reached, no more may be added.
        /// </summary>
        /// <value>The max vertices.</value>
        [Category("Behavior")]
        [Obsolete("Not yet available in GoogleMaps API v3")]
        public int? EditingMaxVertices { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether [enable drawing].
        /// Allows a user to construct a GPolygon object by clicking on additional points on the map. 
        /// The Polygon must already be added to the map, even if the polygon is initially unpopulated and contains no vertices. 
        /// Each click adds an additional vertex to the polygon boundary, and drawing may be terminated through either 
        /// a double-click, or clicking on the first point to complete the shape, at which point an "endline" event will be 
        /// triggered if the polygon was successfully completed; otherwise, a "cancelline" event will be triggered, 
        /// but the polygon will not be removed from the map. 
        /// </summary>
        /// <value><c>true</c> if [enable drawing]; otherwise, <c>false</c>.</value>
        [Category("Behavior")]
        [Obsolete("Not yet available in GoogleMaps API v3")]
        public bool EnableDrawing { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether [enable editing].
        /// Allows modification of an existing GPolygon boundary. 
        /// When enabled, users may select and drag existing vertices. 
        /// The Polygon must already be added to the map via GMap2.addOverlay(). 
        /// Unless a vertex limit less than current number of vertices is specified by maxVertices within EditingOptions, 
        /// "ghost" points will also be added at the midpoints of polygon sections, 
        /// allowing users to interpolate new vertices by clicking and dragging these additional vertices. 
        /// A "lineupdated" event will be triggered whenever vertex is added or moved. 
        /// </summary>
        /// <value><c>true</c> if [enable editing]; otherwise, <c>false</c>.</value>
        [Category("Behavior")]
        [Obsolete("Not yet available in GoogleMaps API v3")]
        public bool EnableEditing { get; set; }

        #endregion
        #endregion

        #region Construct /////////////////////////////////////////////////////////////////////////

        /// <summary>
        /// Initializes a new instance of the <see cref="GooglePolygon"/> class.
        /// </summary>
        public GooglePolygon() {
            this.FillOpacity = .5F;
            this.IsClickable = true;
            this.StrokeOpacity = .5F;
            this.StrokeWeight = 5;
        }
        #endregion

        #region Methods ///////////////////////////////////////////////////////////////////////////
        #endregion
    }
}

File: Artem.GoogleMap/Scripts/GooglePolygon.js
///<reference name="MicrosoftAjax.debug.js"/>
///<reference path="GoogleCommons.js"/>
///<reference path="GoogleMap.js"/>

Type.registerNamespace("Artem.Google");

//#region Polygon class

Artem.Google.Polygon = function Artem_Google_Polygon(map, state) {
    /// <param name="map" type="Artem.Google.Map"></param>
    /// <param name="state" type="Artem.Google.Polygon"></param>

    this.Bounds = state.Bounds;
    this.EditingFromStart = state.EditingFromStart;
    this.EditingMaxVertices = state.EditingMaxVertices;
    this.EnableDrawing = state.EnableDrawing;
    this.EnableEditing = state.EnableEditing;
    this.FillColor = state.FillColor;
    this.FillOpacity = state.FillOpacity;
    this.IsClickable = state.IsClickable;
    this.IsGeodesic = state.IsGeodesic;
    this.Points = state.Points;
    this.StrokeColor = state.StrokeColor;
    this.StrokeOpacity = state.StrokeOpacity
    this.StrokeWeight = state.StrokeWeight;
    this.ZIndex = state.ZIndex;

    this.get_map = function () { return map; };

    this.initialize();
}
Artem.Google.Polygon.prototype = {

    //#region Fields ------------------------------------------------------------------------------

    Bounds: null,
    EditingFromStart: null,
    EditingMaxVertices: null,
    EnableDrawing: null,
    EnableEditing: null,
    FillColor: null,
    FillOpacity: null,
    IsClickable: null,
    IsGeodesic: null,
    Points: null,
    StrokeColor: null,
    StrokeOpacity: null,
    StrokeWeight: null,
    ZIndex: null,

    //#endregion

    //#region Properties --------------------------------------------------------------------------

    get_map: null,
    get_polygon: null,

    //#endregion

    //#region Methods -----------------------------------------------------------------------------

    initialize: function Artem_Google_Polygon$initialize() {

        var options = {
            clickable: this.IsClickable,
            fillColor: null,
            fillOpacity: this.FillOpacity,
            geodesic: this.IsGeodesic,
            map: this.get_map().get_map(),
            paths: null,
            strokeColor: null,
            strokeOpacity: this.StrokeOpacity,
            strokeWeight: this.StrokeWeight,
            zIndex: this.ZIndex
        };

        if (this.FillColor)
            options.fillColor = this.FillColor.HtmlValue;
        if (this.StrokeColor)
            options.strokeColor = this.StrokeColor.HtmlValue;
        if (this.Points) {
            options.paths = new Array();
            var point;
            for (var i = 0; i < this.Points.length; i++) {
                point = this.Points[i];
                options.paths.push(new google.maps.LatLng(point.Latitude, point.Longitude));
            }
        }

        var polygon = new google.maps.Polygon(options);
        this.get_polygon = function () { return polygon; };
    },

    dispose: function Artem_Google_Polygon$dispose() {
        // first dispose the event listeners
        google.maps.event.clearInstanceListeners(this.get_polygon());

        delete this.get_polygon;
        delete this.get_map;

        delete this.Bounds;
        delete this.EditingFromStart;
        delete this.EditingMaxVertices;
        delete this.EnableDrawing;
        delete this.EnableEditing;
        delete this.FillColor;
        delete this.FillOpacity;
        delete this.IsGeodesic;
        delete this.IsClickable;
        delete this.Points;
        delete this.StrokeColor;
        delete this.StrokeOpacity;
        delete this.StrokeWeight;
        delete this.ZIndex;
    },

    save: function Artem_Google_Polygon$save() {
        // TODO - this is not yet implemented/available in GoogleMaps API v3
        //        this.Bounds = this.getBounds();
    },
    //#endregion

    //#region Google Maps API Wrapped -------------------------------------------------------------

    getMap: function Artem_Google_Polygon$getMap() {
        return this.get_polygon().getMap();
    },

    getPath: function Artem_Google_Polygon$getPath() {
        return this.get_polygon().getPath();
    },

    getPaths: function Artem_Google_Polygon$getPaths() {
        return this.get_polygon().getPaths();
    },

    setMap: function Artem_Google_Polygon$setMap(map) {
        this.get_polygon().setMap(map);
    },

    setOptions: function Artem_Google_Polygon$setOptions(options) {
        this.get_polygon().setOptions(options);
    },

    setPath: function Artem_Google_Polygon$setPath(path) {
        this.get_polygon().setPath(path);
    },

    setPaths: function Artem_Google_Polygon$setPaths(paths) {
        this.get_polygon().setPaths(paths);
    }
    //#endregion
}
//#<

Artem.Google.Polygon.registerClass("Artem.Google.Polygon");